[Merged by Bors] - feat(NumberTheory/DirichletCharacter): vanishing of Gauss sums when add char is imprimitive

loefflerd opened the pull request:
The aim of this PR is to show that `gaussSum χ e`, for `χ` a Dirichlet character mod `N` and `e` an additive character mod `N`, is zero if `χ` is primitive but `e` is not. This will be used to prove the functional equation for Dirichlet L-series in a later PR.

Along the way we add various general-purpose lemmas about divisibility in `ZMod N` and other finite rings, and about conductors and primitivity of Dirichlet and additive characters.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

erdOne commented on Mathlib/Algebra/GroupWithZero/NonZeroDivisors.lean:
```diff
@@ -187,6 +187,9 @@ theorem isUnit_of_mem_nonZeroDivisors {G₀ : Type*} [GroupWithZero G₀] {x : G
     inv_mul_cancel (nonZeroDivisors.ne_zero hx)⟩, rfl⟩
 #align is_unit_of_mem_non_zero_divisors isUnit_of_mem_nonZeroDivisors
 
+lemma mem_nonZeroDivisors_of_isUnit {a : M} (ha : IsUnit a) : a ∈ M⁰ :=
```
Maybe `IsUnit.mem_nonZeroDivisors` instead to enable dot notation?

loefflerd commented on Mathlib/NumberTheory/DirichletCharacter/Basic.lean:
```diff
@@ -41,39 +40,53 @@ Main definitions:
 
 lemma eval_modulus_sub (x : ZMod n) : χ (n - x) = χ (-x) := by simp
 
-lemma periodic {m : ℕ} (hm : n ∣ m) : Function.Periodic χ m := by
-  intro a
-  rw [← ZMod.natCast_zmod_eq_zero_iff_dvd] at hm
-  simp only [hm, add_zero]
```
_Note to reviewers of this PR_: I removed this lemma, and the import needed to state it, because it doesn't actually do anything. The Dirichlet character `χ` coerces to a function `ZMod n → R` so `m` is getting coerced into `ZMod n`; and the hypotheses imply that `m` coerces to zero in `ZMod n`. So this is actually showing that `χ` is periodic with period 0, which is not a useful statement. 

(Maybe `Function.Periodic (χ ∘ Int.cast) m` is what was intended, but it is probably better to show that via periodicity of `Int.cast`.)

MichaelStollBayreuth commented on Mathlib/NumberTheory/DirichletCharacter/Basic.lean:
```diff
@@ -96,23 +109,52 @@ lemma changeLevel_eq_cast_of_dvd {m : ℕ} (hm : n ∣ m) (a : Units (ZMod m)) :
 lemma dvd {d : ℕ} (h : FactorsThrough χ d) : d ∣ n := h.1
 
-/-- The Dirichlet character at level `d` through which `χ` factors -/
+/-- The Dirichlet character at level `d` through which `χ` factors. -/
```
Quoting from the [Documentation style](https://leanprover-community.github.io/contribute/doc.html) page:
```quote
 If a doc string is a complete sentence, then it should end in a period.
```
Which I read as "a period is not required when the doc string is not a complete sentence". (But a period is certainly allowed in such a case.)

MichaelStollBayreuth commented on Mathlib/NumberTheory/DirichletCharacter/Basic.lean:
```diff
@@ -96,23 +109,52 @@ lemma changeLevel_eq_cast_of_dvd {m : ℕ} (hm : n ∣ m) (a : Units (ZMod m)) :
+lemma existsUnique {d : ℕ} [NeZero n] (h : FactorsThrough χ d) :
+    ∃! χ' : DirichletCharacter R d, χ = changeLevel h.dvd χ' := by
+  rcases h with ⟨hd, χ₂, rfl⟩
+  exact ⟨χ₂, rfl, fun χ₃ hχ₃↦ (changeLevel_injective hd hχ₃).symm⟩
```
```suggestion
  exact ⟨χ₂, rfl, fun χ₃ hχ₃ ↦ (changeLevel_injective hd hχ₃).symm⟩
```

MichaelStollBayreuth commented on Mathlib/NumberTheory/DirichletCharacter/GaussSum.lean:
```diff
@@ -0,0 +1,77 @@
+    ZMod.intCast_eq_intCast_iff_dvd_sub] at hu
+
+/-- If `gaussSum χ e ≠ 0`, and `d` is such that `e.mulShift d = 1`, then `χ` must factor through
+`d`. (This will be used to show that Gauss sums vanish when `χ` is primitive and `e` is not.)-/
```
```suggestion
`d`. (This will be used to show that Gauss sums vanish when `χ` is primitive and `e` is not.) -/
```

MichaelStollBayreuth commented on Mathlib/NumberTheory/DirichletCharacter/GaussSum.lean:
```diff
@@ -0,0 +1,77 @@
+    gaussSum χ (e.mulShift a) = χ⁻¹ a * gaussSum χ e := by
+  by_cases ha : IsUnit a
+  · conv_rhs => rw [← gaussSum_mulShift χ e ha.unit]
+    have : IsUnit (χ a) := by rw [← ha.unit_spec, ← MulChar.coe_toUnitHom]; apply Units.isUnit
```
Maybe add
```lean
lemma MulChar.map_unit (χ : MulChar R R') {a : R} (ha : IsUnit a) : IsUnit (χ a)
```
(with suitable `R` and `R'`) in `Mathlib.NumberTheory.LegendreSymbol.MulCharacter`? 
It looks like a generally useful statement.

loefflerd replied:
In the course of attempting to add it, I realised someone had already done so – it's `IsUnit.map`.

MichaelStollBayreuth commented on Mathlib/NumberTheory/DirichletCharacter/GaussSum.lean:
```diff
@@ -0,0 +1,77 @@
+    rw [IsUnit.unit_spec, MulChar.inv_apply_eq_inv, Ring.inverse_mul_cancel_left _ _ this]
+  · rw [MulChar.map_nonunit _ ha, zero_mul]
+    apply gaussSum_eq_zero_of_isPrimitive_of_not_isPrimitive _ hχ
+    -- this is just showing `e.mulShift a` is not primitive if `a` is non-unit -- separate?
```
I think it makes sense to have a separate lemma for this.

MichaelStollBayreuth replied:
How does it relate to `not_isPrimitive_mulShift` below?

loefflerd replied:
Good point – I separated that out as a separate lemma and then forgot to update the original proof using it.

loefflerd commented:
Thanks Michael for the review! I have updated the PR accordingly. 

In the process, I realised that I had (coincidentally) added 2 of the 3 lemmas suggested in a "TODO" note at the top of `DirichletCharacter/Basic.lean`. So I have completed the job by adding the third TODO lemma (`changeLevel_eq_one_iff`) to this PR, and removed the note.

MichaelStollBayreuth commented on Mathlib/Algebra/GroupWithZero/NonZeroDivisors.lean:
```diff
@@ -242,6 +245,15 @@ theorem nonZeroDivisors_le_comap_nonZeroDivisors_of_injective [NoZeroDivisors M'
 #align non_zero_divisors_le_comap_non_zero_divisors_of_injective nonZeroDivisors_le_comap_nonZeroDivisors_of_injective
 
+/-- In a finite ring, an element is a unit iff it is a non-zero-divisor. -/
+lemma isUnit_iff_mem_nonZeroDivisors_of_finite [Finite R] {a : R} :
+    IsUnit a ↔ a ∈ nonZeroDivisors R := by
```
It is good to have that! I'm surprised it wasn't there already...

MichaelStollBayreuth reviewed (approved):
LGTM!

MichaelStollBayreuth commented:
maintainer merge

riccardobrasca commented:
Thanks!

bors merge
